internal/controller/finalizer/types.go

```diff
@@ -0,0 +1,98 @@
+// Copyright Red Hat, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the License);
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an AS IS BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package finalizer
+
+import (
+	"context"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	"github.com/openshift-service-mesh/federation/internal/controller"
+)
+
+// FinalizeFn is a closure which can be used in the reconciler to define finalizer logic just before
+// the object is removed from kube-apiserver.
+type FinalizeFn func() error
+
+// Handler encapsulates finalizer handling. It can:
+// - add finalizer to a given object and persist it
+// - perform cleanup defined as FinalizeFn if the object is marked for deletion
+//
+// Example usage in the controller reconcile loop:
+//
+//	finalizerHandler := finalizer.NewHandler(r.Client, "federation.openshift-service-mesh.io/mesh-federation")
+//	if finalized, errFinalize := finalizerHandler.Finalize(ctx, meshFederation, func() error {
+//		// finalizer logic
+//		return nil
+//	}); finalized || errFinalize != nil {
+//		return ctrl.Result{}, errFinalize
+//	}
+//
+//	if justAdded, errAdd := finalizerHandler.Add(ctx, meshFederation); justAdded || errAdd != nil {
+//		return ctrl.Result{}, errAdd
+//	}
```

Comment:
Change related to other two comments.

```suggestion
//	finalizerHandler := finalizer.NewHandler(r.Client, "federation.openshift-service-mesh.io/mesh-federation")
//	if finalizeNecessary, errFinalize := finalizerHandler.Finalize(ctx, meshFederation, func() error {
//		// finalizer logic
//		return nil
//	}); finalizeNecessary {
//		return ctrl.Result{}, errFinalize
//	}
//
//	if finalizerPresent, err := finalizerHandler.Add(ctx, meshFederation); !finalizerPresent {
//		return ctrl.Result{}, errAdd
//	}
```